boot/worker/src/boot/pgp.clj

```diff
@@ -2,56 +2,74 @@
-
-(defn get-pubkey
-  [secring & [user]]
-  (let [ks      (pgp/list-public-keys secring)
-        re-user #(re-find (re-pattern (Pattern/quote (str user))) %)
-        match?  #(some re-user (-> % pgp/key-info :user-ids))]
-    (let [ks' (if-not user ks (filter match? ks))]
-      (if (or (not user) (<= (count ks') 1))
-        (first ks')
-        (throw (ex-info "multiple keys match"
-                 {:keys (mapv pgp/key-info ks')}))))))
-
-(defn secret-key-for
-  [secring pubkey]
-  (pgp/get-secret-key secring (pgp/key-id pubkey)))
-
-(defn prompt-for
-  [keyring user-id]
-  (if-let [pk (-> keyring secring (get-pubkey user-id))]
-    (-> pk pgp/key-info :user-ids first (str "\nGPG passphrase: "))
-    (throw (Exception. (format "public key not found (%s)" user-id)))))
-
-(defn sign
-  [content passphrase & {:keys [keyring user-id]}]
-  (let [sr (secring keyring)
-        pk (get-pubkey sr user-id)
-        sk (secret-key-for sr pk)
-        pr (pgp/unlock-key sk passphrase)]
-    (-> content (pgp/sign :sha1 pr) pgp/encode-ascii)))
+(defn find-secring []
+  (let [home (System/getenv "HOME")
+        locations [(System/getenv "GNUPGHOME") (System/getenv "BOOT_GPG_SECRING") (str home "/.gnupg/secring.gpg") (str home "/.gpg/secring.gpg")]
+        exists-fn #(when (and % (.exists (io/as-file %))) %)
+        secring (some exists-fn locations)]
+    (or secring (throw (Exception. "Can't find secring file. Please set 'BOOT_GPG_SECRING'")))))
+
+(defn get-keyring []
+  (keyring/load-secret-keyring (io/file (find-secring))))
+
+(defn find-master-key-by-user [user]
+  (let [keyring (get-keyring)
+        pub-keys (keyring/list-public-keys keyring)
+        re-user #(re-find (re-pattern user) %)
+        match?  #(some re-user (-> % pgp/key-info :user-ids))
+        key (some #(when (match? %) %) pub-keys)]
+    (or key (throw (Exception. (format "User %s not found" user))))))
+
+(defn find-master-key []
+  (if-let [user (System/getenv "BOOT_GPG_USER")]
+    (find-master-key-by-user user)
+    (let [keyring (get-keyring)
+          master-keys (filter (comp :master-key? pgp/key-info) (keyring/list-public-keys keyring))]
+      (cond
+        (= 0 (count master-keys)) (throw (Exception. "No master key found"))
+        (= 1 (count master-keys)) (first master-keys)
+        :else (throw (Exception. "Multiple master keys found. Please set 'BOOT_GPG_USER'"))))))
+
+(defn get-private-key-by-id [hex-id passphrase]
+  (let [keyring (get-keyring)
+        seckey (keyring/get-secret-key keyring hex-id)]
+    (pgp/unlock-key seckey passphrase)))
+
+(defn get-private-key [passphrase]
+  (let [keyring (get-keyring)]
+    (if-let [hex-id (System/getenv "BOOT_GPG_SIGNING_KEY_ID")]
+      (get-private-key-by-id hex-id passphrase)
+      (-> (find-master-key)
+          pgp/key-info
+          :key-id
+          (#(keyring/get-secret-key keyring %))
+          (pgp/unlock-key passphrase)))))
+
+(defn read-pass
```

Comment:
I know, lein uses `gpg-agent` which I quite like as an approach because it shifts complexity on the tool itself that you can configure using `.gpgconf`. I'd rather have `push` accept optionally the passphrase so in my CI machines I can have a `profile.boot` that does `(task-options! push {:passphrase "My pass"})`. Environmental variable might be overkill in this case.